pkg/scheduler/metrics/resources/resources.go

```diff
@@ -0,0 +1,237 @@
+				case v1.ResourceCPU:
+					unitName = "cores"
+				case v1.ResourceMemory:
+					unitName = "bytes"
+				case v1.ResourceStorage:
+					unitName = "bytes"
+				case v1.ResourceEphemeralStorage:
+					unitName = "bytes"
+				default:
+					switch {
+					case v1helper.IsHugePageResourceName(resourceName):
+						unitName = "bytes"
+					case v1helper.IsAttachableVolumeResourceName(resourceName):
+						unitName = "integer"
+					}
+				}
+				var priority string
+				if p.Spec.Priority != nil {
+					priority = strconv.FormatInt(int64(*p.Spec.Priority), 10)
+				}
+				recordMetricWithUnit(ch, t.desc.total, p.Namespace, p.Name, p.Spec.NodeName, p.Spec.SchedulerName, priority, resourceName, unitName, val)
+			}
+		}
+	}
+}
+
+func recordMetricWithUnit(
+	ch chan<- metrics.Metric,
+	desc *metrics.Desc,
+	namespace, name, nodeName, schedulerName, priority string,
+	resourceName v1.ResourceName,
+	unit string,
+	val resource.Quantity,
+) {
+	if val.IsZero() {
+		return
+	}
+	ch <- metrics.NewLazyConstMetric(desc, metrics.GaugeValue,
+		val.AsApproximateFloat64(),
+		namespace, name, nodeName, schedulerName, priority, string(resourceName), unit,
+	)
+}
+
+// podRequestsAndLimitsByLifecycle returns a dictionary of all defined resources summed up for all
+// containers of the pod. If PodOverhead feature is enabled, pod overhead is added to the
+// total container resource requests and to the total container limits which have a
+// non-zero quantity. The caller may avoid allocations of resource lists by passing
+// a requests and limits list to the function, which will be cleared before use.
+// This method is the same as v1resource.PodRequestsAndLimits but avoids allocating in several
+// scenarios for efficiency.
+func podRequestsAndLimitsByLifecycle(pod *v1.Pod, reuseReqs, reuseLimits v1.ResourceList) (reqs, limits v1.ResourceList, terminal bool) {
+	switch {
+	case len(pod.Spec.NodeName) == 0:
+		// unscheduled pods cannot be terminal
+	case pod.Status.Phase == v1.PodSucceeded, pod.Status.Phase == v1.PodFailed:
+		terminal = true
+		// TODO: add a condition for deleted pods where all containers are terminated once scheduler
+		//   has the same logic
+	}
+	if terminal {
+		return
+	}
+
+	// attempt to reuse the maps if passed, or allocate otherwise
+	reqs, limits = reuseOrClearResourceList(reuseReqs), reuseOrClearResourceList(reuseLimits)
+
+	for _, container := range pod.Spec.Containers {
+		v1resource.AddResourceList(reqs, container.Resources.Requests)
+		v1resource.AddResourceList(limits, container.Resources.Limits)
+	}
+	// init containers define the minimum of any resource
+	for _, container := range pod.Spec.InitContainers {
+		v1resource.MaxResourceList(reqs, container.Resources.Requests)
+		v1resource.MaxResourceList(limits, container.Resources.Limits)
+	}
+
+	// if PodOverhead feature is supported, add overhead for running a pod
+	// to the sum of reqeuests and to non-zero limits:
+	if pod.Spec.Overhead != nil {
```

Comment:
Would be good to guard this additionally by feature gate: (also be consistent with other places)

https://github.com/kubernetes/kubernetes/blob/d41f79154251aaa54120015d8ad5b69319b44af8/pkg/scheduler/framework/types.go#L623-L624

Reply:
Weird, I thought I copied that out of core library.  Will verify.

Reply:
That's why we should have a single implementation :)

Reply:
I don't think it was in the core library when i copied it :)